raft-tech/TANF-app
# Front/429/upload

### This pull request changes...

- Add `reports/[selected year]/upload` route to application
- Can successfully upload a TXT, XLS, or CSV file to S3
- Add error handling for unsuccessful upload to each specific input


## FUNCTIONALITY
*Describe the functionality*
This allows a user to be able to upload files to S3. Upon navigation to `Reports` a user can select a year and then click `Begin Report` which will take them to the upload page. From there, the user can upload a file for each of the four categories.

**NOTE**
The `Submit Report` button is not functioning in this PR but will be addressed in #430.

Just a small reminder, before this is ready for official approval there are backend tests that need to be fixed @carltonsmith @riatzukiza 

*Addresses Issue #*
#429 

## TO TEST

1. Log into application
2. Click on `Reports` tab
3. Select a year
4. Click `Begin Report` button
5. Click on input to upload a file / drag and drop a file into the input
6. Go to `Network` tab in dev tools and you should see the following successful requests

<img width="884" alt="Screen Shot 2021-02-01 at 11 25 46 AM" src="https://user-images.githubusercontent.com/23202538/106507676-3d2ec980-6480-11eb-8a55-bc07f7f7bb4b.png">

There is a successful call to `signed_url` which is our backend API endpoint to connect to S3
There are two successful calls to S3 - one is the preflight check to make sure we are authorized and the second is the actual successful `PUT` to S3.

To test a failure:

Follow steps 1 - 4. 
5. Open devtools and set network online throttle to `offline`.
6. Try and upload file through selection or drag and drop
7. A `Network Error` should pop up with styling changes on the specific input that had trouble.


### This pull request is ready to merge when...

  - OpenAPI  
  - Readme
  - Entity Relationship Diagram (ERD)
